// @rahsheen/react-wizard/dist/cjs/components/Wizard.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = __importStar(require("react"));
var useWizard_1 = __importDefault(require("../hooks/useWizard"));
var Step_1 = __importDefault(require("./Step"));
var WizardContext = React.createContext(null);
var Wizard = function (props) {
    var children = props.children, steps = props.steps, initialValues = props.initialValues;
    if (!(children || steps))
        return null;
    var enabledSteps = React.Children.toArray([steps, children]).filter(function (child) { return !child.props.disabled; });
    var _a = useWizard_1.default({
        initialValues: initialValues,
        size: enabledSteps.length,
        onSubmit: props.onSubmit
    }), nextStep = _a.nextStep, prevStep = _a.prevStep, index = _a.index, onChangeValue = _a.onChangeValue, onSubmit = _a.onSubmit, values = _a.values;
    var value = React.useMemo(function () { return ({
        currentIndex: index,
        prevStep: prevStep,
        nextStep: nextStep,
        onChangeValue: onChangeValue,
        onSubmit: onSubmit,
        values: values,
        isLast: index === enabledSteps.length - 1
    }); }, [index]);
    var activeStep = enabledSteps[index];
    return Boolean(activeStep) ? (React.createElement(WizardContext.Provider, { value: value }, activeStep)) : null;
};
function useWizardContext() {
    var context = React.useContext(WizardContext);
    if (!context) {
        throw new Error("Wizard compound components cannot be rendered outside the Wizard component");
    }
    return context;
}
Wizard.Step = function (props) { return React.createElement(Step_1.default, __assign({}, useWizardContext(), props)); };
exports.default = Wizard;
