// @rahsheen/react-wizard/dist/esm/hooks/useWizard.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { useState } from "react";
function useWizard(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.initialValues, initialValues = _c === void 0 ? {} : _c, _d = _b.size, size = _d === void 0 ? 2 : _d, _e = _b.onSubmit, onSubmitWizard = _e === void 0 ? null : _e;
    var _f = useState(0), index = _f[0], setIndex = _f[1];
    var _g = useState(__assign({}, initialValues)), values = _g[0], setValues = _g[1];
    var onSubmit = function () {
        onSubmitWizard(values);
    };
    var prevStep = function () {
        if (index <= 0)
            return;
        setIndex(index - 1);
    };
    var nextStep = function () {
        if (index >= size - 1)
            return;
        setIndex(index + 1);
    };
    var onChangeValue = function (name, value) {
        setValues(function (oldValues) {
            var _a;
            return (__assign({}, oldValues, (_a = {}, _a[name] = value, _a)));
        });
    };
    return {
        nextStep: nextStep,
        prevStep: prevStep,
        index: index,
        onChangeValue: onChangeValue,
        onSubmit: onSubmit,
        values: values
    };
}
export default useWizard;
